### PR TITLE
Add Looker SDK and streaming support

### PR DESCRIPTION
**Pull request by tjbanghart**
Adds functionality described in [go/scaling-avatica-looker-api](http://goto.google.com/scaling-avatica-looker-api).

To enable the the Looker SDK use `jdbc:looker` (as opposed to `jdbc:avatica:remote`) as the JDBC URL protocol.
NOTE: This is awaiting changes from [b/288031194](https://b.corp.google.com/issues/288031194). Don't expect this to work properly without that work completed.

**TODOs:**
- [x] Unit tests for _everything_ but definitely the streaming parser.
- [x] Use `sql_interface` endpoints.
- [x] Ensure OAuth tokens are set as `AuthToken`s rather than raw header.
- [ ] Handle `metadata` from `json_bi` result format. We currently skip right to `rows`.
- [x] Support prepared statements. i.e. those that `Prepare` then later `Execute`. Currently we only override `PrepareAndExecute`
- [ ] https://github.com/looker-open-source/sdk-codegen/issues/1341 - not going to happen for awhile. Requires large refactor to the SDK gen.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+    if (responseCode == 200) {
+      // grab the input stream and write to the output for the main thread to consume.
+      InputStream inputStream = connection.getInputStream();
+      int bytesRead;
+      byte[] buffer = new byte[DEFAULT_STREAM_BUFFER_SIZE];
+      while ((bytesRead = inputStream.read(buffer)) != -1) {
+        outputStream.write(buffer, 0, bytesRead);
+      }
+      outputStream.close();
+      inputStream.close();
```
Comment (barrkel) [inline core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java]: This `close` looks like it's going to race with the read side. From my scan of the source of PipedInput/OutputStream implementation, `PipedOutputStream` called `PipedInputStream.receive` which waits for space if necessary but otherwise writes to a circular buffer. If bytesRead is small enough to fit in the buffer (1024 bytes by default), this can close the stream before the read side reads it out.

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+        throw new RuntimeException(e);
+      }
+    });
+    Thread.UncaughtExceptionHandler exceptionHandler = (th, ex) -> {
+      try {
+        in.close();
+        out.close();
+        throw new RuntimeException(ex);
+      } catch (IOException e) {
+        throw new RuntimeException(e);
```
Comment (barrkel) [inline core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java]: The documented JVM behavior for exceptions thrown from the uncaught exception handler is to ignore them. You do get output on the terminal. The client code in `LookerIterable` is just going to get a pipe closed exception though.

Depending on how much implementation detail information you want to transmit from this thread to the client, it may be worthwhile marshaling exceptions from the streaming thread to the thread consuming the iterable, by e.g. sharing an atomic reference between the two, storing an exception before closing the pipe, and checking the atomic upon getting a pipe closed exception on the other side.

Of course the [JDK doesn't help](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/io/PipedInputStream.java#L266C50-L266C50) with a specific exception type for pipe closed, it's just a custom message inside an `IOException`. There are a few ways to solve this problem, exercise for the reader...

File: core/src/main/java/org/apache/calcite/avatica/MetaImpl.java [hunk 3]
```diff
@@ -1527,12 +1527,12 @@ public Object next() {
 
   /** Iterable that yields an iterator over rows coming from a sequence of
    * {@link Meta.Frame}s. */
-  private class FetchIterable implements Iterable<Object> {
```
Comment (barrkel) [inline core/src/main/java/org/apache/calcite/avatica/MetaImpl.java]: Changing the visibility on these implementation details suggests to me they should probably be top-level classes.

File: core/src/main/java/org/apache/calcite/avatica/remote/Driver.java [hunk 4]
```diff
@@ -59,7 +63,8 @@ public Driver() {
    */
   public enum Serialization {
     JSON,
-    PROTOBUF;
+    PROTOBUF,
+    LOOKER;
```
Comment (barrkel) [inline core/src/main/java/org/apache/calcite/avatica/remote/Driver.java]: I agree with the comment below, that serialization doesn't feel quite right here. You're replacing the endpoint, not just the wire format. I'm not sure what alternative to suggest though as it probably requires bigger refactoring to encapsulate the logic of pluggable back ends. Whether it's worth it or not depends on whether this code is expected to get upstreamed, and I don't know the intentions here.

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+    // Set the auth header as the SDK would
+    connection.setRequestProperty("Authorization",
+        "token " + authSession.getAuthToken().getAccessToken());
+    connection.setRequestProperty("Accept", "application/json");
+    int responseCode = connection.getResponseCode();
+    if (responseCode == 200) {
+      // grab the input stream and write to the output for the main thread to consume.
+      InputStream inputStream = connection.getInputStream();
+      int bytesRead;
+      byte[] buffer = new byte[DEFAULT_STREAM_BUFFER_SIZE];
```
Comment (barrkel) [inline core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java]: This buffer size doesn't control the actual buffer (byte window of I/O concurrency) between the HTTP connection and the client iterating over `LookerIterable`. You're not guaranteed to get a full buffer-worth from HTTP / TCP transport, and the pipe only has a tiny buffer of 1024 by default. If the client is slow at iterating, it may have to wait for network transport every time. That may be an acceptable design, I don't know what the design parameters are around slow clients.

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+   *  Add streaming support to the Kotlin SDK.
+   */
+  private void streamResponse(String url, OutputStream outputStream) throws IOException {
+    // use some SDK client helpers to tighten up this call
+    AuthSession authSession = getSdk().getAuthSession();
+    Transport sdkTransport = authSession.getTransport();
+    // makes a proper URL from the API endpoint path as the SDK would.
+    String endpoint = sdkTransport.makeUrl(url, Collections.emptyMap(), null);
+    URL httpsUrl = new URL(endpoint);
+    HttpsURLConnection connection = (HttpsURLConnection) httpsUrl.openConnection();
```
Comment (barrkel) [inline core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java:172]: You probably want to set some timeouts here, or transport them from configuration somehow. What is the latency for connect, for reads from Looker, and so on. Not really expected to have actual values in code this early but it needs coordination for reasonable values on both sides, and if read latency is allowed to be high then you probably want more buffering.

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+
+    @Override
+    public Iterator<Object> iterator() {
+      return new LookerIterator(statement, state, firstFrame);
+    }
+
+    /**
+     * Iterator that parses an {@link InputStream} of JSON into a sequence of {@link Meta.Frame}s.
+     */
+    public class LookerIterator extends FetchIterator implements Iterator<Object> {
```
Comment (barrkel) [inline core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java]: Have you considered running the parsing logic in the thread connected to Looker, and using a blocking queue of Frame objects (or a very similar envelope) as the buffer to invert control, converting Supplier-oriented (push) into Iterator-oriented (pull)? This structure could solve a number of problems:

1. If the queue uses an envelope, it could contain three types of objects: a bundle of rows, an exception (for transmitting those Looker connection exceptions) or an End Of Stream terminal object (eliminating the race on close).
1. The computational cost of parsing JSON and constructing rows is taken out of the client read path, reducing latency.
1. The queue's length (indirectly) controls the amount of buffering. This is slightly double-edged though, since large or wide frames may mean the queue retains reference to unexpectedly large object graphs.

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java (hunk 6 above)
Comment (tjbanghart) [reply core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java:172]: Ah of course, we set a timeout for the SDK Transport that should also be applied here. Thank you for catching this.

File: core/src/main/java/org/apache/calcite/avatica/remote/Driver.java (hunk 4 above)
Comment (tjbanghart) [reply core/src/main/java/org/apache/calcite/avatica/remote/Driver.java]: https://github.com/looker-open-source/calcite-avatica/pull/2/commits/761d6d64f1cc011906cfd4b0d0c426ecb145778e removes this `serialization` bit and adds a new Looker-specific `org.apache.calcite.avatica.remote.looker.Driver` implementation. This can be used via `jdbc:looker` as the URL protocol without changing the behavior of `jdbc:avatica:remote` connections.

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java (hunk 6 above)
Comment (tjbanghart) [reply core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java:172]: Added in https://github.com/looker-open-source/calcite-avatica/pull/2/commits/b407594dc18df89006383cd8abedce40d761a336

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java (hunk 7 above)
Comment (tjbanghart) [reply core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java]: This sounds way more desirable than leaving the pipes as they are. 

So, the Looker HTTP thread would parse complete frames up to the client configured`fetchSize` (e.g. 100 rows at a time). Once a frame is complete, it is enqueued on the `BlockingQueue`. If the queue is full, the thread blocks until a frame is read off the queue.

The client would then dequeue complete frames, blocking if no frame is available. It no longer has to worry about generating these frames. 

If we are worried about point 3. we could limit the `fetchSize` to something we feel comfortable with - or even compute a max frame size based on column width + row length requested. Perhaps we can revisit that part after some testing.

File: core/src/main/java/org/apache/calcite/avatica/MetaImpl.java [hunk 8]
```diff
@@ -1590,7 +1590,14 @@ public Object next() {
       return o;
     }
 
-    private void moveNext() {
+    /** A helper method to call {@link Meta#fetch}. Pulled out from {@link #moveNext()} so that
+     * extending classes can override the fetch implementation without calling out to a Meta. */
+    protected Frame doFetch(StatementHandle h, long currentOffset, int fetchSize)
+        throws NoSuchStatementException, MissingResultsException {
+      return fetch(h, currentOffset, fetchSize);
+    }
```
Comment (wnob) [inline core/src/main/java/org/apache/calcite/avatica/MetaImpl.java]: What's the problem with allowing subclasses to simply override `fetch`?

File: build.gradle.kts [hunk 9]
```diff
@@ -176,6 +176,8 @@ allprojects {
     plugins.withId("java-library") {
         dependencies {
             "implementation"(platform(project(":bom")))
+            // Add the locally bundled LookerSDK fat jar
+            "implementation"(files("../libs/looker-kotlin-sdk-f91f8ad.jar"))
```
Comment (wnob) [inline build.gradle.kts]: Feels like we should file a cleanup bug to make this integration easier and throw it in the backlog for now.

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/Driver.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+import org.apache.calcite.avatica.remote.Service;
+import org.apache.calcite.avatica.remote.looker.utils.LookerSdkFactory;
+
+import com.looker.sdk.LookerSDK;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.Properties;
+
+public class Driver extends UnregisteredDriver {
```
Comment (wnob) [inline core/src/main/java/org/apache/calcite/avatica/remote/looker/Driver.java]: This class needs a javadoc.

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/Driver.java [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    connection.setService(service);
+    return new LookerRemoteMeta(connection, service);
+  }
+
+  @Override public Connection connect(String url, Properties info)
+      throws SQLException {
+    AvaticaConnection conn = (AvaticaConnection) super.connect(url, info);
+
+    if (conn == null) {
+      // It's not an url for our driver
```
Comment (wnob) [inline core/src/main/java/org/apache/calcite/avatica/remote/looker/Driver.java]: nit: Would a better comment be something like "the URL did not match Looker's JDBC connection string prefix"?

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/utils/package-info.java [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Static utilities for {@link org.apache.calcite.avatica.remote.looker}
+ */
+package org.apache.calcite.avatica.remote.looker.utils;
```
Comment (wnob) [inline core/src/main/java/org/apache/calcite/avatica/remote/looker/utils/package-info.java]: http://go/no-utils

I would just move `LookerSdkFactory` up a package level, into the `...remote.looker` package.

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java [hunk 13]
```diff
@@ -0,0 +1,380 @@
...
+      connection.setHostnameVerifier(trustAllHostNames);
+    } catch (NoSuchAlgorithmException | KeyManagementException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * A regrettable necessity. The Kotlin SDK relies on an outdated Ktor HTTP client based on Kotlin
+   * Coroutines which are a tad difficult to work with in Java. Here we make a HTTP client to handle
+   * the response stream ourselves which is "good enough" for now. This method should be revisited
```
Comment (wnob) [inline core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java]: "good enough" is hard to parse. Better to spell out the downside.

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java [hunk 14]
```diff
@@ -0,0 +1,380 @@
...
+   * when the Kotlin SDK has built-in streams.
+   *
+   * TODO https://github.com/looker-open-source/sdk-codegen/issues/1341:
+   *  Add streaming support to the Kotlin SDK.
+   */
+  private void streamResponse(String url, OutputStream outputStream) throws IOException {
+    // use some SDK client helpers to tighten up this call
```
Comment (wnob) [inline core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java]: "tighten up this call" is also a bit hard to parse. Is this a TODO?

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/utils/LookerSdkFactory.java [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+
+  /**
+   * Makes the API endpoint to run a previously made query.
+   */
+  public static String queryEndpoint(Long id) {
+    return String.format(Locale.ROOT, QUERY_ENDPOINT, TransportKt.encodeParam(id), RESULT_FORMAT);
+  }
+
+  /**
+   * Makes the SDK call and throws any errors as runtime {@link SQLException}s
```
Comment (wnob) [inline core/src/main/java/org/apache/calcite/avatica/remote/looker/utils/LookerSdkFactory.java]: This javadoc is confusing. Looks like it would not throw a `SQLException`.

File: core/src/test/java/org/apache/calcite/avatica/remote/looker/DriverTest.java [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+
+  @Test
+  public void lookerDriverIsRegistered() throws SQLException {
+    Driver driver = DriverManager.getDriver("jdbc:looker:url=foobar.com");
+    assertThat(driver, is(instanceOf(org.apache.calcite.avatica.remote.looker.Driver.class)));
+  }
+
+  @Test
+  public void driverThrowsAuthExceptionForBlankProperties() throws SQLException {
+    Properties props = new Properties();
```
Comment (wnob) [inline core/src/test/java/org/apache/calcite/avatica/remote/looker/DriverTest.java]: http://go/unit-test-practices#structure

I find the arrange - act - assert structure does a lot for test readability. Just put some blank space in between each section. It's OK if some tests are more like arrange - act - assert - act - assert.

**Review by wnob — COMMENTED** [review]
Leaving some general readability comments for now. Will think harder about this later. Since this is going to be a permanent "fixup" commit, we should be pretty careful about making sure it remains easy to rebase on master for every bump, so good to think about exactly how to break it into smaller commits for easy maintenance in posterity.

File: build.gradle.kts (hunk 9 above)
Comment (tjbanghart) [reply build.gradle.kts]: Yeah definitely - we will be building this jars locally for the foreseeable future. There was talk of getting this on maven and tied to official releases but I don't think that will happen soon.

File: core/src/main/java/org/apache/calcite/avatica/MetaImpl.java (hunk 8 above)
Comment (tjbanghart) [reply core/src/main/java/org/apache/calcite/avatica/MetaImpl.java]: That's a good question. It made sense at the time. but ~~I'd rather not change this so we can leave `MetaImpl` alone.~~ Strange wording on my part. Meant to say: We should revert this change so that `doFetch` is removed and `Meta#fetch` is used. That way we don't have to touch `MetaImpl`. We can override `fetch` in `LookerMetaImpl`.

**Comment by tjbanghart** [comment]
Thanks for the review @wnob! 

I am still exploring what we would need to have upstream, but my hope is that we can keep all meaningful changes in `org.apache.calcite.avatica.remote.looker`. I would like this to be an extension of the remote Avatica driver rather than a fixup to existing implementations. There might be some changes here that others would benefit from, but I don't believe much of this should make it into main with the exception of some class + method visibility changes (which do need Jira cases).

That said, response streaming into a series of `Frame`s may be something to lift out of this and would probably benefit most from open-source contributions. Perhaps it makes sense to open a Jira case for this as well... on the other hand the streamed response is specific to Looker's `json_bi` format.

IMO, Avatica has been open enough that we have not had to modify existing classes extensively. Any areas where this PR has done such modifications highlight potential Jira cases for upstream.

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java (hunk 7 above)
Comment (tjbanghart) [reply core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java]: aa93bd90a48bc55d54238d13ca883aa1bd0443c0 makes the change described above using `FrameEnvelope`. The "End Of Stream terminal object" is a `Frame` object with a `done` value of `true`.

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/utils/package-info.java (hunk 12 above)
Comment (tjbanghart) [reply core/src/main/java/org/apache/calcite/avatica/remote/looker/utils/package-info.java]: Done

File: core/src/main/java/org/apache/calcite/avatica/MetaImpl.java (hunk 8 above)
Comment (tjbanghart) [reply core/src/main/java/org/apache/calcite/avatica/MetaImpl.java]: Removed this method so we no longer touch `MetaImpl`. Making `RemoteMeta` `public` and its constructor `protected` are now the only changes outside of the `remote.looker` package.

Well, besides bundling the sdk and adding the looker driver  to `META-INF/services/java.sql.Driver`

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java [hunk 17]
```diff
@@ -0,0 +1,438 @@
...
+      new ConcurrentHashMap();
+
+  /**
+   * Wrapper for either a {@link Frame} or {@link Exception}. Allows for {@link #stmtQueueMap} to
+   * hold complete Frames and present exceptions when consumers are ready to encounter them.
+   */
+  private class FrameEnvelope {
+
+    final Frame frame;
+    final Exception exception;
```
Comment (barrkel) [inline core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java]: Reading this class in isolation, I wonder what it means if both frame and exception are null, or both are not null?

What do you think about representing this as a sum type so that it works well with future (Java 17+) pattern matching (or well with Kotlin today)? It's a smidgen more awkward at the use point when lacking pattern matching, but it's probably a cleaner representation than this one, which has 4 combinations of null/not-null in construction.

Like how we might do a sum types in Kotlin:
```
  sealed class FrameEnvelope {
    class OK(val frame: Frame) : FrameEnvelope()
    class Error(val exception: Exception) : FrameEnvelope()
  }
```

Or in Java:
```
  class FrameEnvelope {
    private FrameEnvelope() {}
    public static final class OK extends FrameEnvelope { ... }
    public static final class Error extends FrameEnvelope { ... }
  }
```

I do feel myself that it's a bit academic, but when looking at the definition of a type, I appreciate it when invalid states cannot be represented.

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java [hunk 18]
```diff
@@ -0,0 +1,438 @@
...
+    default:
+      throw new IOException("Unable to parse " + columnMetaData.type.rep + " from stream!");
+    }
+  }
+
+  @Override
+  public Frame fetch(final StatementHandle h, final long offset, final int fetchMaxRowCount)
+      throws NoSuchStatementException, MissingResultsException {
+    // If this statement was initiated as a LookerFrame then it will have an entry in the queue map
+    if (stmtQueueMap.containsKey(h)) {
```
Comment (barrkel) [inline core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java]: Just an observation really: the implementation of this method is sensitive to being called correctly - it must be invoked sequentially, not concurrently (`containsKey() followed by get()` is not atomic), and it must not be invoked after receiving a frame with the `done` flag set - but writing a defensive implementation seems to come with other downsides, like retaining references to handles.

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java [hunk 19]
```diff
@@ -0,0 +1,438 @@
...
+   * for more details.
+   */
+  private final ConcurrentMap<StatementHandle, BlockingQueue<FrameEnvelope>> stmtQueueMap =
+      new ConcurrentHashMap();
+
+  /**
+   * Wrapper for either a {@link Frame} or {@link Exception}. Allows for {@link #stmtQueueMap} to
+   * hold complete Frames and present exceptions when consumers are ready to encounter them.
+   */
+  private class FrameEnvelope {
```
Comment (barrkel) [inline core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java]: Should this be `private static class` (i.e. are you missing **static**)?

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java [hunk 20]
```diff
@@ -0,0 +1,438 @@
...
+
+            rows.add(row);
+            rowsRead++;
+          }
+          // we fetched the allowed number of rows so add the complete frame to the queue
+          currentOffset += rowsRead;
+          frameQueue.put(makeFrame(currentOffset, /*done=*/false, rows));
+        }
+      } catch (Exception e) {
+        // enqueue all exceptions for the main thread to report
```
Comment (barrkel) [inline core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java]: The comment says `exceptions` but I only see one? To be clear I think failing fast on first exception makes sense, so the comment is probably wrong.

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java [hunk 21]
```diff
@@ -0,0 +1,438 @@
...
+  private LookerSDK getSdk() {
+    return lookerService.sdk;
+  }
+
+  /**
+   * A single meta can have multiple running statements. This map keeps track of
+   * {@code FrameEnvelopes}s that belong to a running statement. See {@link #prepareStreamingThread}
+   * for more details.
+   */
+  private final ConcurrentMap<StatementHandle, BlockingQueue<FrameEnvelope>> stmtQueueMap =
```
Comment (barrkel) [inline core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java]: I wondered if `StatementHandle` was the right thing here. It comes with a `Signature` attached, which looks like it could be sizeable, and the generated Protobuf class will do a value based equality on it. It's probably be OK, but worth bearing in mind - it appears to contain all column metadata, the SQL, and a bunch of parameters. Every access could be recursively comparing every field, if the handles for successive invocations have different instances (e.g. they come from different RPC invocations / different protobuf deserializations) - I don't know the call pattern.

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java [hunk 22]
```diff
@@ -0,0 +1,438 @@
...
+      queue.put(makeException(e));
+    } catch (InterruptedException ex) {
+      throw new RuntimeException(ex);
+    }
+  }
+
+  /**
+   * Prepares a thread to stream a query response into a series of {@link FrameEnvelope}s.
+   */
+  protected Thread prepareStreamingThread(String baseUrl, Signature signature, int fetchSize,
```
Comment (barrkel) [inline core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java]: How about breaking this and some of the associated parsing logic into a separate class, and leaving `LookerRemoteMeta` as a thing which spawns threads and connects them to queues? The class is a bit big now and it's not clear which code runs on the caller's thread and which code runs in a streaming thread.

The stream parser itself probably doesn't need to know about threads, and if it takes a `Consumer<FrameEnvelope>` argument it needn't know about queues either. The value of the consumer argument could be a lambda wrapping `blockingQueue::put` that handles `InterruptedException`.

Then if someone wants to test the parser in isolation, they don't need to mess around with threads or queues.

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java [hunk 23]
```diff
@@ -0,0 +1,438 @@
...
+
+  private void seekToValue(JsonParser parser) throws IOException {
+    while (parser.nextToken() != null && !VALUE_KEY.equals(parser.currentName())) {
+      // seeking to `value` key for the field e.g. `"rows": [{"field_1": {"value": 123 }}]
+    }
+    // now move to the actual value
+    parser.nextToken();
+  }
+
+  private void putExceptionOrFail(BlockingQueue queue, Exception e) {
```
Comment (barrkel) [inline core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java]: I guess `BlockingQueue` started out polymorphic, containing either `Frame` or `Exception` instances, but we shouldn't use raw generic collections, it should be updated to `BlockingQueue<FrameEnvelope>` throughout. It would eliminate some explicit casts too.

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java (hunk 18 above)
Comment (tjbanghart) [reply core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java]: Discussed over GVC

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java (hunk 17 above)
Comment (tjbanghart) [reply core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java]: I split the difference here and made the constructor private - `LookerFrameEnvelope` has two static factory methods `ok(Frame)` and `error(Exception)`. There was an issue with static members of inner classes but I hope what is there now is close enough (if not happy to change it).

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java (hunk 19 above)
Comment (tjbanghart) [reply core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java]: Moved to its own class

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java (hunk 21 above)
Comment (tjbanghart) [reply core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java]: Changed to statement id (int) to match the rest of Avatica.

File: core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java (hunk 22 above)
Comment (tjbanghart) [reply core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java]: Moved into `LookerResponseParser` class and added some tests.

**Review by barrkel — APPROVED** [review]
